typed/helpers.go

```diff
@@ -217,7 +217,9 @@ func keyedAssociativeListItemToPathElement(a value.Allocator, s *schema.Schema,
 		} else if def != nil {
 			keyMap = append(keyMap, value.Field{Name: fieldName, Value: value.NewValueInterface(def)})
 		} else {
-			return pe, fmt.Errorf("associative list with keys has an element that omits key field %q (and doesn't have default value)", fieldName)
+			// Don't add the key to the key field list.
+			// A key field list where it is set then represents a different entry
+			// in the associate list.
 		}
 	}
 	keyMap.Sort()
```

Comment:
might be a good idea to make sure we can't end up with a completely empty keyMap by doing a check like this after the loop:

```
	if len(list.Keys) > 0 && len(keyMap) == 0 {
		return pe, fmt.Errorf("associative list with keys has an element that omits all key fields %q (and doesn't have default values for any key fields)", list.Keys)
	}
```

Reply:
Pushed.